keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -115,15 +122,31 @@ A few possible semantics are described here:
 
 #### Equality
 
-Equality of two fields is very important:
-- Scalar fields are compared directly: fields and values must exactly match from old
-  object to new object.
-- Lists fields are compared using the same logic currently used for native
-  kubernetes types: Empty and null lists are equal. A missing list is NOT equal
-  to a null or empty list.
-- Same rule for list applies to maps and struct: An empty map and a null map are
-  considered equal. The keys in the map must be the same. Values are compared
-  using this same set of rules.
+Equality is strict, i.e. using `reflect.DeepEqual`.
+
+This has the following consequences for
+
+- **native resources:** these go through an unmarshal step from proto or JSON into the 
+native Golang struct. This normalizes undefined, null and empty lists, maps and 
+structs to `nil`. That struct is converted to JSON before the immutability check.
```

Comment:
nit: Roundtripping through proto (as happens in etcd) normalizes null and empty lists. JSON preserves the difference between undefined, null, and empty.

Reply:
doesn't proto preserve the distinction between nil and empty maps?

Reply:
nil and empty will lead to no item of the repeated type in proto AFAIK. All three are identified.

Reply:
This bit isn't accurate for JSON:
> This normalizes undefined, null and empty lists, maps and structs to `nil`

JSON API requests differentiate between undefined/null and empty for lists and maps

Reply:
not talking about the requests, but `json.Unmarshal` semantics. That normalizes, and then `json.Marshal` then gives the same result (we don't care which actually).

Reply:
> not talking about the requests, but json.Unmarshal semantics. That normalizes

json.Unmarshal does not normalize null and []/{}

Reply:
Would it make sense to grid it out? E.g.

Before:

| | undefined | null | empty list | empty map | empty struct |
| --- | --- | --- | --- | --- | --- |
| undefined | ? | ? | ? | ? |
| null | ? | ? | ? | ? |
| empty list | ? | ? | ? | ? |
| empty map | ? | ? | ? | ? |
| empty struct | ? | ? | ? | ? |

After:

| | undefined | null | empty list | empty map | empty struct |
| --- | --- | --- | --- | --- | --- |
| undefined | == | | | |
| null | | == | | |
| empty list | | | == | | |
| empty map | | | | == | |
| empty struct | | | | | == |

Reply:
I do love a good table

Reply:
that would definitely help. the tricky bit is when the client (or multiple clients) and server are using different mechanisms that silently coerce some of those categories. For example:
1. typed CR client generated from go types, speaking JSON to the API server
    * loses distinction between null and absent
    * always omits empty (non-struct/non-custom) values when serializing (for omitempty fields)
    * always includes empty values when serializing (for non-omitempty fields)
2. unstructured CR client (like kubectl), speaking JSON to the API server
    * preserves distinction between null and absent
    * does not have omitempty info available, so preserves input perfectly
3. API server, serving CRs (behaves like an unstructured client)

Reply:
Gotcha. Lots of moving parts.